### PR TITLE
gapic: generate mix-in API methods on client libraries

### PR DESCRIPTION
**Pull request by noahdietz**
This adds support for client library mixin method generation. The (draft) AIP is at https://github.com/aip-dev/google.aip.dev/pull/694 and has the high level context. Ping me for the internal doc(s) if desired.

The supported mixin services are:
- `google.longrunning.Operations`
- `google.iam.v1.IAMPolicy`
- `google.cloud.location.Locations`

The proto descriptors for these services are loaded from their compiled Go types, fed into `pbinfo` for lookup availability, and included in generation of each product service client. The mixin API gRPC stubs (i.e. `IAMPolicyClient`) are added to the gapic client struct as an unexported field and initialized with the connection established with the product service endpoint (`i.e. pubsub.googleapis.com`). The mixin methods get all of the GAPIC features - auto-pagination, client metadata, default deadline & retry settings (if configured), request header params, client option support, etc.

Preview of Showcase [client regeneration](https://github.com/googleapis/gapic-showcase/pull/649) with these changes (as of first review request).

### CODE REVIEW

File: internal/gengapic/mixins.go [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+func getLocationsDescriptor() *descriptor.FileDescriptorProto {
+	return mixinFiles["google.cloud.location.Locations"][0]
+}
+
+func getLocationsMethods() []*descriptor.MethodDescriptorProto {
+	return getLocationsDescriptor().GetService()[0].GetMethod()
+}
+
+func getIAMPolicyDescriptors() []*descriptor.FileDescriptorProto {
+	return mixinFiles["google.iam.v1.IAMPolicy"]
```
Comment (codyoss) [inline internal/gengapic/mixins.go:105]: nit: add `[0]` for consistency with other getters

**Review by codyoss — APPROVED** [review]
LGTM

File: internal/gengapic/client_init.go [hunk 2]
```diff
@@ -138,10 +142,10 @@ func (g *generator) clientOptions(serv *descriptor.ServiceDescriptorProto, servN
 func (g *generator) clientInit(serv *descriptor.ServiceDescriptorProto, servName string) error {
 	p := g.printf
 
-	var hasLRO bool
+	var hasLRORpc bool
```
Comment (tbpg) [inline internal/gengapic/client_init.go:145]: It's not great, but this should be `hasLRORPC`. Perhaps `hasRPCForLRO`?

https://github.com/golang/go/wiki/CodeReviewComments#initialisms

File: internal/gengapic/client_init_test.go [hunk 3]
```diff
@@ -147,13 +151,34 @@ func TestClientOpt(t *testing.T) {
 		t.Fatal(err)
 	}
 
+	servIamOverride := &descriptor.ServiceDescriptorProto{
```
Comment (tbpg) [inline internal/gengapic/client_init_test.go:154]: IAM

File: internal/gengapic/mixins.go (hunk 1 above)
Comment (noahdietz) [reply internal/gengapic/mixins.go:105]: Ah that's the thing, this one returns multiple proto files, while the others are just one proto file.

**Review by tbpg — APPROVED** [review]
A couple super minor comments.

File: internal/gengapic/client_init.go (hunk 2 above)
Comment (noahdietz) [reply internal/gengapic/client_init.go:145]: sgtm! Done.

File: internal/gengapic/client_init_test.go (hunk 3 above)
Comment (noahdietz) [reply internal/gengapic/client_init_test.go]: Done.

File: internal/gengapic/mixins.go (hunk 1 above)
Comment (codyoss) [reply internal/gengapic/mixins.go:105]: Is this function used outside this file? What I was trying to say I guess is the other methods all just do return a single item. Noticed that this is the only one that returned a list. The get method calls in the rest just operate on the descriptor directly([0] was already called when getting descriptor)

Comment (noahdietz) [reply internal/gengapic/mixins.go:105]: It's not used outside this file, no. I see what you are saying, yeah it's a little weird. I'll change it up.

Comment (noahdietz) [reply internal/gengapic/mixins.go]: Done! Thanks for the explanation, Cody.

**Comment by noahdietz** [comment]
I confirmed locally that regeneration of google-cloud-go had no unexpected changes - the config files haven't been updated to enable this feature yet.

**Pull request merged**